assets/scripts/program-favorite.js

```diff
@@ -0,0 +1,280 @@
+ */
+export default class FavoritePrograms {
+    docReady() {
+        // Bail if browser does not support localStorage.
+        if ( typeof Storage === 'undefined' ) {
+            return;
+        }
+
+        this.cache();
+        this.favorites = this.parseFavorites();
+
+        if ( ! this.hasFavorites() ) {
+            this.showNoResults();
+        }
+
+        // Init functionalities common to all templates.
+        this.initCommon();
+
+        // Init functionalities for favorite-listing.
+        this.initFavorites();
+
+        // Favorite popup functionalities
+        this.favoritePopup();
+    }
+
+    cache() {
+        this.classes = {
+            hasFavorites: 'has-favorites',
+            isFavorite: 'is-favorite',
+            hide: 'is-hidden',
+        };
+
+        this.selectors = {
+            button: '.js-toggle-favorite',
+        };
+        this.storageKey = 'tredu-program-favorites';
+        this.closeFavorites = $( '#js-favorites-close' );
+        this.openFavorites = $( '#js-favorites-open' );
+        this.favoritesContainer = $( '#js-favorites-container' );
+        this.favoritePrograms = $( '#js-favorite-programs' );
+        this.favoritesContent = $( '#js-favorites-content' );
+        this.favoritesNoResults = $( '#js-favorites-no-results' );
+    }
+
+    /**
+     * Initialize functionalities common for all templates.
+     */
+    initCommon() {
+        this.favoriteButtons = $( this.selectors.button );
+
+        if ( this.favoriteButtons.length > 0 ) {
+            this.populateFavorites();
+        }
+    }
+
+    /**
+     * Initialize functionalities for favorites listing.
+     */
+    initFavorites() {
+        // Fetch favorites according to localStorage.
+        if ( this.favorites.length !== 0 ) {
+            dp( 'ProgramFavorites/FavoritePrograms', {
+                partial: 'header-favorites-item',
+                tidy: true,
+                data: true,
+                args: {
+                    ids: this.favorites,
+                },
+            } ).then( ( response ) => {
+                if ( ! response.success ) {
+                    this.showNoResults();
+                    return;
+                }
+
+                // Set favorites content.
+                this.favoritePrograms.html( response.success );
+
+                // Show the container.
+                this.favoritePrograms.removeClass( this.classes.hide );
+            } ).catch( ( error ) => {
+                console.log( error );
+                this.showNoResults();
+            } );
+        }
+
+        // Update favorites when favorite-button clicked
+        if ( this.favoritesContainer.length > 0 ) {
+            this.favoriteButtons.on( 'click', ( e ) => {
+                e.stopPropagation();
+                this.toggleFavorite( e.currentTarget );
+                this.loadFavorites();
+            } );
+
+            $( document ).on( 'click', '.js-remove-favorite', ( e ) => {
```

Comment:
Voisiko tällä classilla tehdä myös jonkun `this.removeFavoriteButtons`in ja tehdä event listenerin samalla tavalla kuin tossa yllä?

Reply:
Piti toteuttaa tälleen, koska `.js-remove-favorite` tulee dynaamisesti listaukseen yksittäisen suosikki-itemin mukana